### DISCUSSION TITLE
How to make libCurl use hexadecimal cnonce?

### DISCUSSION THREAD

**Discussion by ganeshkamath89** [curl/curl#6399] [Q&A] [answered]
I  was working on a project with WinHTTP on Windows where the cnonce was hexadecimal.

I have explained my problem further in this stack overflow question: 
[https://stackoverflow.com/questions/65488754/how-to-make-libcurl-use-hexadecimal-cnonce-instead-of-alphanumeric-in-digest-aut](https://stackoverflow.com/questions/65488754/how-to-make-libcurl-use-hexadecimal-cnonce-instead-of-alphanumeric-in-digest-aut)

I want to know if there is some Curl Opt header that can help me generate hexadecimal nonce instead of the default alphanumeric nonce.

**Comment by jay** [comment]
> I want to know if there is some Curl Opt header that can help me generate hexadecimal nonce instead of the default alphanumeric nonce.

There's no option to do that. IIRC the cnonce generated by libcurl is a base64 encoding of random hex digits.

**Reply by ganeshkamath89** [reply]
what is IIRC, is it a library you use to generate cnonce?

**Reply by jay** [reply]
**I**f **I** **R**ecall **C**orrectly

**Reply by ganeshkamath89** [reply]
Thanks for providing the solution to my problem below 👍

**Comment by bagder** [comment, marked as answer]
There's no such option. It wouldn't make any sense to provide one.

curl is written to work with Digest as defined in the RFC documents. RFC 7616 even suggests that "[It is advised that this string be Base64 or hexadecimal data](https://tools.ietf.org/html/rfc7616#section-3.3)".

It sounds as if your server has a broken Digest implementation if that causes any problems. And btw, curl has done this for a very long time and seems to be inter-operating pretty good with most servers like this.

**Reply by ganeshkamath89** [reply]
Hi Bagder. Like you have mentioned, it can be either base64 or hexadecimal. But right now Curl is giving Base64 value and there is no option to get hexadecimal value.

The client nonce (cnonce) is generated by the client. Our earlier client (WinHTTP) gave hexadecimal and keeping this as reference the server application was written for hexadecimal cnonce. The server application developers have since moved on to other projects. So we need some mechanism to change the cnonce used otherwise our client-server communication will fail.

So I am asking you if some kind of header / switch can be provided to use hexadecimal value with libcurl for client side cnonce.

**Reply by jay** [reply]
We don't have a reason to believe base64 cnonce is a problem, nobody has reported that so far. libcurl is open source so you could of course modify it yourself:

https://github.com/curl/curl/blob/e052859759b34d0e05ce0f17244873e5cd7b457b/lib/vauth/digest.c#L710-L723

~~~diff
diff --git a/lib/vauth/digest.c b/lib/vauth/digest.c
index 559852f..d01ab4f 100644
--- a/lib/vauth/digest.c
+++ b/lib/vauth/digest.c
@@ -699,8 +699,6 @@ static CURLcode auth_create_digest_http_message(
   unsigned char ha1[65];    /* 64 digits and 1 zero byte */
   unsigned char ha2[65];    /* 64 digits and 1 zero byte */
   char userh[65];
-  char *cnonce = NULL;
-  size_t cnonce_sz = 0;
   char *userp_quoted;
   char *response = NULL;
   char *hashthis = NULL;
@@ -710,18 +708,15 @@ static CURLcode auth_create_digest_http_message(
     digest->nc = 1;
 
   if(!digest->cnonce) {
-    char cnoncebuf[33];
-    result = Curl_rand_hex(data, (unsigned char *)cnoncebuf,
-                           sizeof(cnoncebuf));
-    if(result)
-      return result;
+    size_t cnonce_sz = 33;
 
-    result = Curl_base64_encode(data, cnoncebuf, strlen(cnoncebuf),
-                                &cnonce, &cnonce_sz);
+    digest->cnonce = malloc(cnonce_sz);
+    if(!digest->cnonce)
+      return CURLE_OUT_OF_MEMORY;
+
+    result = Curl_rand_hex(data, (unsigned char *)digest->cnonce, cnonce_sz);
     if(result)
       return result;
-
-    digest->cnonce = cnonce;
   }
 
   if(digest->userhash) {
~~~

**Reply by ganeshkamath89** [reply]
Thank you very much. This is extremely helpful.

**Comment by bagder** [comment]
> right now Curl is giving Base64 value and there is no option to get hexadecimal value.

... right, and my point was that the server is violating the spec by not accepting that value so I hope you report this issue to them rather than being happy with painting over the problem by just changing curl!

**Reply by ganeshkamath89** [reply]
Hi Bagder. We report problems very frequently, but the solution to problems is either given very late or delayed so long that the project itself concludes without the fix. So as a fall back, I now have a mechanism to make it work because I just need to replace the code for now. From our side we have already reported the problem.
